libcaf_core/caf/actor_system.hpp

```diff
@@ -109,6 +109,39 @@ class CAF_CORE_EXPORT actor_system {
   actor_system(const actor_system&) = delete;
   actor_system& operator=(const actor_system&) = delete;
 
+  /// Calls a cleanup function in its destructor for cleaning up global state.
+  class [[nodiscard]] global_state_guard {
+  public:
+    using void_fun_t = void (*)();
+
+    explicit global_state_guard(void_fun_t f) : fun_(f) {
+      // nop
+    }
+
+    global_state_guard(global_state_guard&& other) noexcept : fun_(other.fun_) {
+      other.fun_ = nullptr;
+    }
+
+    global_state_guard& operator=(global_state_guard&& other) noexcept {
+      std::swap(fun_, other.fun_);
+      return *this;
+    }
```

Comment:
I'm wondering if this could create a scenario where the move constructor will not call whatever was in `fun_` before, but the move assignment will. One swaps while the other one leaves `other.fun_` with a `nullptr`, thus discarding whatever `fun_` was previously?

Reply:
The move constructor just transfers ownership. In the case of the move assignment operator, the operation should generally be equivalent to destroying `this` and then "re-create" it with the content of `other`.

So, when move-assigning something to a non-null guard, both cleanup functions should eventually be called. The two common ways to do that are by either just swapping with `other` or move `other` into a temporary first and then swapping. The latter makes sure that we release whatever resource is managed before leaving the operator. Since this guard is managing some global state, my thought was that it doesn't really matter when the calls to the cleanup functions happen as long as they do get called. So calling `swap` here seemed like the most efficient way to go about things.